File: product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        /*
                    [4,5,1,8,2]
            L=>   [1,4,20,20,160]
                   ^
            R=>   [80,16,16,2,1]
                               ^
                               
            Result=> [L] * [R]
            
        */
        int N = nums.length;
        
        int[] result = new int[N];
        
        int[] L = new int[N];
        L[0] = 1;
        
        int[] R = new int[N];
        R[N - 1] = 1;
        
        // compute product of everything to the left of nums[i]
        for (int i = 1; i < N; i++) {
            L[i] = L[i - 1] * nums[i - 1];
        }
        
        // compute product of everything to the right of nums[i]
        for (int i = N - 2; i >= 0; i--) {
            R[i] = R[i + 1] * nums[i + 1];
        }
        
        // store the result to result array
        for (int i = 0; i < N; i++) {
            result[i] = L[i] * R[i];
        }
        
        return result;
    }
}

File: product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        
        left = [1] * len(nums)
        right = [1] * len(nums)
        ans = [1] * len(nums)
        
        for i in range(1, n):
            left[i] = left[i - 1] * nums[i - 1]
        
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]
            
        for i in range(0, n):
            ans[i] = left[i] * right[i];
            
        return ans